### PR TITLE
Nanogpt

### PR DESCRIPTION
**Pull request by ezhang7423**
## Type of pull request:
- [ ] Bug fix
- [x] Feature
- [ ] Other

## Description:
- Wrote a homemade engine wrapper with the nanogpt player and the weight file from the training loop in chess-nanoGPT. Implemented the search function using the evaluation code in chess_gpt_eval to output the model's predicted best move given a current board/game state.

## Related Issues:

PR: https://github.com/teamcore-harvard/chess-nanoGPT/pull/2

### CODE REVIEW

File: NanoGPT_Engine.py
Comment (ezhang7423) [inline NanoGPT_Engine.py:1]: change thie filename to be NanoGPT_Enginer.py -> nanogpt_engine.py. this is the standard in python

File: lib/engine_wrapper.py [hunk 1]
```diff
@@ -60,7 +60,12 @@ def create_engine(engine_config: config.Configuration, game: Optional[model.Game
     elif engine_type == "uci":
         Engine = UCIEngine
     elif engine_type == "homemade":
-        Engine = getHomemadeEngine(cfg.name)
+        # Engine = getHomemadeEngine(cfg.name)
+        from NanoGPT_Engine import NanoGPTEngine
+        # Engine = NanoGPTEngine(os.environ["WEIGHT_FILE"], cfg.temperature) #TODO: Add the temperature hyperparameter to the config file (set to 0.)
+        options = remove_managed_options(cfg.lookup(f"{engine_type}_options") or config.Configuration({}))
+        logger.debug(f"Starting engine: {commands}")
+        return NanoGPTEngine(os.environ["WEIGHT_FILE"], commands, options, stderr, cfg.draw_or_resign, game, cwd=cfg.working_dir, temperature=cfg.temperature) #TODO: Add the temperature hyperparameter to the config file (set to 0.)
```
Comment (ezhang7423) [inline lib/engine_wrapper.py]: finish this todo, and add it to the argparser instead of the config.yml

File: lib/matchmaking.py [hunk 2]
```diff
@@ -52,7 +52,8 @@ def __init__(self, li: LICHESS_TYPE, config: Configuration, user_profile: USER_P
         self.matchmaking_cfg = config.matchmaking
         self.user_profile = user_profile
         self.last_challenge_created_delay = Timer(seconds(25))  # Challenges expire after 20 seconds.
-        self.last_game_ended_delay = Timer(minutes(self.matchmaking_cfg.challenge_timeout))
+        # self.last_game_ended_delay = Timer(minutes(self.matchmaking_cfg.challenge_timeout))
+        self.last_game_ended_delay = Timer(seconds(1))
```
Comment (ezhang7423) [inline lib/matchmaking.py]: add a comment: # to speed up matchmaking

File: lib/matchmaking.py [hunk 3]
```diff
@@ -89,7 +90,7 @@ def should_create_challenge(self) -> bool: #!!!
     def create_challenge(self, username: str, base_time: int, increment: int, days: int, variant: str,
                          mode: str) -> str:
         """Create a challenge."""
-        params = {"rated": mode == "rated", "variant": variant}
+        params = {"rated": mode == "rated", "variant": variant, "color": 'white' }
```
Comment (ezhang7423) [inline lib/matchmaking.py]: remove "color": 'white"

File: lichess-bot.py [hunk 4]
```diff
@@ -45,7 +45,8 @@
 
 logger = logging.getLogger(__name__)
 
-with open("lib/versioning.yml") as version_file:
+# with open("lib/versioning.yml") as version_file:
+with open("/mnt/data/lichess_bot_eval_part_2/lichess-bot/lib/versioning.yml") as version_file:
```
Comment (ezhang7423) [inline lichess-bot.py]: revert back to the old one? if the old one is not working, then use the absolute path `Path('lib/versioning.yml').absolute()`

File: lichess-bot.py [hunk 5]
```diff
@@ -366,7 +367,7 @@ def lichess_bot_main(li: LICHESS_TYPE,
                                              challenge_queue,
                                              play_game_args,
                                              active_games,
-                                             max_games)
+                                             max_games) #What is the point of the correspondence game
```
Comment (ezhang7423) [inline lichess-bot.py]: remove

File: lichess-bot.py [hunk 6]
```diff
@@ -510,6 +511,7 @@ def game_error_handler(error: BaseException) -> None:
     pool.apply_async(play_game,
                      kwds=play_game_args,
                      error_callback=game_error_handler)
+    # play_game(**play_game_args)
```
Comment (ezhang7423) [inline lichess-bot.py:519]: this seems important - why was it commented?

File: lichess-bot.py [hunk 7]
```diff
@@ -572,8 +574,8 @@ def handle_challenge(event: EVENT_TYPE, li: LICHESS_TYPE, challenge_queue: MULTI
         li.decline_challenge(chlng.id, reason=decline_reason)
 
 
-@backoff.on_exception(backoff.expo, BaseException, max_time=600, giveup=lichess.is_final,  # type: ignore[arg-type]
-                      on_backoff=lichess.backoff_handler)
+# @backoff.on_exception(backoff.expo, BaseException, max_time=600, giveup=lichess.is_final,  # type: ignore[arg-type]
```
Comment (ezhang7423) [inline lichess-bot.py]: uncomment

Comment (ezhang7423) [reply lichess-bot.py]: also add a comment that this should be commented out when debugging

File: lichess-bot.py [hunk 8]
```diff
@@ -1041,16 +1043,25 @@ def start_lichess_bot() -> None:
     parser.add_argument("--config", help="Specify a configuration file (defaults to ./config.yml).")
     parser.add_argument("-l", "--logfile", help="Record all console output to a log file.", default=None)
     parser.add_argument("--disable_auto_logging", action="store_true", help="Disable automatic logging.")
+    # parser.add_argument("--iter_num", help="Describes which iteration of the model training we are on.")
```
Comment (ezhang7423) [inline lichess-bot.py]: del this line

File: lichess-bot.py [hunk 9]
```diff
@@ -1041,16 +1043,25 @@ def start_lichess_bot() -> None:
...
+    parser.add_argument("--weight_file", help="Describes which iteration of the model training we are on.")
     args = parser.parse_args()
 
     logging_level = logging.DEBUG if args.v else logging.INFO
     auto_log_filename = None
     if not args.disable_auto_logging:
-        auto_log_filename = "./lichess_bot_auto_logs/recent.log"
+        # auto_log_filename = "./lichess_bot_auto_logs/recent.log"
+        auto_log_filename = "/mnt/data/lichess_bot_eval_part_2/lichess-bot/lichess_bot_auto_logs/recent.log"
```
Comment (ezhang7423) [inline lichess-bot.py]: again, don't use absolute paths. what if this was running on a  different machine?

File: lichess-bot.py [hunk 10]
```diff
@@ -1041,16 +1043,25 @@ def start_lichess_bot() -> None:
...
     logging_configurer(logging_level, args.logfile, auto_log_filename, True)
     logger.info(intro(), extra={"highlighter": None})
 
-    CONFIG = load_config(args.config or "./config.yml")
+    # CONFIG = load_config(args.config or "./config.yml")
+    CONFIG = load_config(args.config or "/mnt/data/lichess_bot_eval_part_2/lichess-bot/config.yml")
```
Comment (ezhang7423) [inline lichess-bot.py]: fix abspath

File: lichess-bot.py [hunk 11]
```diff
@@ -1041,16 +1043,25 @@ def start_lichess_bot() -> None:
...
+    #this is the weight file. The default is the stockfish executable
+    os.environ["WEIGHT_FILE"] = args.weight_file
+    # CONFIG.engine.working_dir = f"../chess-nanogpt/out/ckpt_{iter_num}.pt" 
+    # CONFIG.engine.dir = "/mnt/data/lichess_bot_eval_part_2/chess-nanogpt/out"
+    # CONFIG.engine.name = f"ckpt_{iter_num}.pt"
```
Comment (ezhang7423) [inline lichess-bot.py]: clean this up

File: nanogpt/nanogpt_module.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+"""
+import os
+import pickle
+from contextlib import nullcontext
+import torch
+import tiktoken
+from nanogpt.model import GPTConfig, GPT
+
+# BASE_DIR = "nanogpt/"
+BASE_DIR = "/mnt/data/lichess_bot_eval_part_2/lichess-bot/nanogpt/"
```
Comment (ezhang7423) [inline nanogpt/nanogpt_module.py]: again

File: nanogpt/nanogpt_module.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+        num_samples = 1  # number of samples to draw
+        max_new_tokens = 6  # number of tokens generated in each sample
+        temperature = 0.01  # 1.0 = no change, < 1.0 = less random, > 1.0 = more random, in predictions
+        top_k = 200  # retain only the top_k most likely tokens, clamp others to have 0 probability
+        seed = 1337
+        # device = "cuda"  # examples: 'cpu', 'cuda', 'cuda:0', 'cuda:1', etc.
+        device = "cpu"
+        dtype = "float16"  # 'float32' or 'bfloat16' or 'float16'
+        compile = False  # use PyTorch 2.0 to compile the model to be faster
+        exec(
```
Comment (ezhang7423) [inline nanogpt/nanogpt_module.py]: delete this. also delete the configurator.py file

File: NanoGPT_Engine.py [hunk 14]
```diff
@@ -0,0 +1,201 @@
+"""
+Some example classes for people who want to create a homemade bot.
+
+With these classes, bot makers will not have to implement the UCI or XBoard interfaces themselves.
+"""
+
+from __future__ import annotations
+import re
```
Comment (ezhang7423) [inline NanoGPT_Engine.py]: remove unused import

File: NanoGPT_Engine.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+# logger.debug("message") will only print "message" if verbose logging is enabled.
+logger = logging.getLogger(__name__)
+
+
+# Return is (move_san, move_uci, attempts, is_resignation, is_illegal_move)
+def get_legal_move(
+    player: NanoGptPlayer,
+    board: chess.Board,
+    game_state: str,
+    player_one: bool = False,
```
Comment (ezhang7423) [inline NanoGPT_Engine.py]: remove. this is not used

File: NanoGPT_Engine.py [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    max_attempts: int = 5,
+# ) -> LegalMoveResponse:
+) -> PlayResult:
+    """Request a move from the player and ensure it's legal."""
+    move_san = None
+    move_uci = None
+
+
+    for attempt in range(max_attempts):
+        move_san = player.get_move(game_state, 0.5)
```
Comment (ezhang7423) [inline NanoGPT_Engine.py]: this temperature is hardcoded at 0.5. make this configurable. see below comment on passing in from argparser

File: NanoGPT_Engine.py [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+        # player.get_move(
+        #     board, game_state, min(((attempt / max_attempts) * 1) + 0.001, temperature)
+        # )
+
+        # Sometimes when GPT thinks it's the end of the game, it will just output the result
+        # Like "1-0". If so, this really isn't an illegal move, so we'll add a check for that.
+        if move_san is not None:
+            if move_san == "1-0" or move_san == "0-1" or move_san == "1/2-1/2":
+                print(f"{move_san}, player has resigned")
+                # TODO
```
Comment (ezhang7423) [inline NanoGPT_Engine.py]: this needs to be done

Comment (ezhang7423) [reply NanoGPT_Engine.py]: or at least add a comment that it doesn't break the bot, and why it doesn't break anything

File: NanoGPT_Engine.py [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+                                        online_moves_cfg,
+                                        draw_or_resign_cfg)
+
+        if isinstance(best_move, list) or best_move.move is None:
+            draw_offered = check_for_draw_offer(game)
+
+            time_limit, can_ponder = move_time(board, game, can_ponder,
+                                               setup_timer, move_overhead,
+                                               is_correspondence, correspondence_move_time)
+
```
Comment (ezhang7423) [inline NanoGPT_Engine.py]: all of this should be deleted

**Comment by ezhang7423** [comment]
vincent new updates: 

Description:
Addressed comments on the nanogpt branch. Added in an argparser line for temperature and setting it as an environment variable. Implemented the resignation if block in get_legal_move by returning a move object with the resigned field as True. See related issues for the full list of comments addressed.

added back the pool.apply_sync function for the play_move function call. This caused a bug with a board form a previous unfinished game being passed in, causing the prior_board to not be initialized properly. Fixed it by raising a chess.engine.EngineError when the board is not an initial board and prior_board is a None type

File: lib/engine_wrapper.py [hunk 19]
```diff
@@ -60,7 +60,12 @@ def create_engine(engine_config: config.Configuration, game: Optional[model.Game
     elif engine_type == "uci":
         Engine = UCIEngine
     elif engine_type == "homemade":
-        Engine = getHomemadeEngine(cfg.name)
+        # Engine = getHomemadeEngine(cfg.name)
+        from nanogpt_engine import NanoGPTEngine
+        # Engine = NanoGPTEngine(os.environ["WEIGHT_FILE"], cfg.temperature) #TODO: Add the temperature hyperparameter to the config file (set to 0.)
```
Comment (ezhang7423) [inline lib/engine_wrapper.py]: remove line 63 and 65

File: lichess-bot.py (hunk 6 above)
Comment (vincentlinzhu) [reply lichess-bot.py:519]: Originally, the line 516 was not there. It was to stop running multiple processes and just run one game at a time. The pool.apply_sync was the original run for play_game that runs multiple games at the same time.

File: nanogpt/nanogpt_module.py (hunk 13 above)
Comment (vincentlinzhu) [reply nanogpt/nanogpt_module.py]: I deleted the exec line and the configurator file. Is that alright? Isn't the configurator supposed to be executed in the nanogpt module.

File: lichess-bot.py (hunk 6 above)
Comment (ezhang7423) [reply lichess-bot.py:519]: OK, this is fine then

**Comment by ezhang7423** [comment]
@vincentlinzhu Congratulations!!

**Pull request merged**